src/ContainerTrait.php

```diff
@@ -14,22 +39,152 @@ public function _unique_element($desired = null)
         return $desired.($postfix > 1 ? ('_'.$postfix) : '');
     }
 
-    protected function _unique() {
+    /**
+     * Extension to add() method which will perform linking of
+     * the object with the current class.
+     */
+    protected function _add_Container($element, $args = [])
+    {
+        // Normalize the arguments, bring name out
+        if (is_string($args)) {
+
+            // passed as string
+            $args=[$args];
+        } elseif (!is_array($args)) {
+
+            throw new Exception(['Second argument must be array','arg2'=>$args]);
+        } elseif (is_array($args) && isset($args['name'])) {
+
+            // passed as ['name'=>'foo'];
+            $args[0]=$args['name'];
+            unset($args['name']);
+        } elseif (isset($element->short_name)) {
+
+            // element has a name already
+            $args[0]=$element->short_name;
+        } elseif (isset($element->_trackableTrait)) {
+
+            // ask element on his preferred name, then make it unique.
+            $cn = $element->getDesiredName();
+            $args[0] = $this->_unique_element($cn);
+        } else {
+
+            // generate name based on the class
+            $cn = str_replace('\\', '_', strtolower(get_class($element)));
+            $args[0] = $this->_unique_element($cn);
+        }
+
+        // Maybe element already exists
+        if (isset($this->elements[$args[0]])) {
+            throw new Exception([
+                'Element with requested name already exists',
+                'element'=>$element,
+                'name'=>$args[0],
+                'this'=>$this,
+                'arg2'=>$args
+            ]);
+        }
+
+        if(isset($this->_appScope) && isset($element->_appScope)) {
+            $element->app = $this->app;
+        }
+
+        $element->owner = $this;
+        $element->short_name = $args[0];
+
+
+        if(isset($element->_trackableTrait)) {
+
+            $element->name = $this->_shorten($this->name.'_'.$element->short_name);
+
+            $this->elements[$element->short_name] = $element;
+        } else {
+            // dont store extra reference to models and controlers
+            // for purposes of better garbage collection.
+            $this->elements[$element->short_name] = true;
+        }
+        
+        return $element;
     }
 
     /**
-     * Extension to add() method which will perform container tracking
+     * Remove child element if it exists.
+     *
+     * @param string $short_name short name of the element
+     *
+     * @return $this
      */
-    protected function _add_Container() {
-    }
+    public function removeElement($short_name)
+    {
 
-    protected function _shorten () {
+        if (is_object($short_name)) {
+            $short_name = $short_name->short_name;
+        }
+        unset($this->elements[$short_name]);
```

Comment:
I remember some strange case when unset() doesn't properly free object from memory.
Shouldn't we do this better?
```
$this->elements[$short_name] = null;
unset($this->elements[$short_name]);
```